moodle/Sniffs/Commenting/FileExpectedTagsSniff.php

```diff
@@ -0,0 +1,160 @@
+     *
+     * @param File $phpcsFile The file being scanned.
+     * @param int $stackPtr The position in the stack.
+     */
+    public function process(File $phpcsFile, $stackPtr) {
+        // Get the stack pointer for the file-level docblock.
+        $stackPtr = Docblocks::getDocBlockPointer($phpcsFile, $stackPtr);
+        if ($stackPtr === null) {
+            // There is no file-level docblock.
+
+            if (TokenUtil::countGlobalScopesInFile($phpcsFile) > 1) {
+                // There are more than one item in the global scope.
+                // Only accept the file docblock.
+                return;
+            } else {
+                // There is only one item in the global scope.
+                // We can accept the file docblock or the item docblock.
+                $stackPtr = $phpcsFile->findNext(Collections::closedScopes(), 0);
+            }
+        }
+
+        $this->processFileCopyright($phpcsFile, $stackPtr);
+        $this->processFileLicense($phpcsFile, $stackPtr);
+    }
+
+    /**
+     * Process the file docblock and check for the presence of a @copyright tag.
+     *
+     * @param File $phpcsFile The file being scanned.
+     * @param int $stackPtr The position in the stack.
+     */
+    private function processFileCopyright(File $phpcsFile, $stackPtr): void {
+        $copyrightTokens = Docblocks::getMatchingDocTags($phpcsFile, $stackPtr, '@copyright');
+        if (empty($copyrightTokens)) {
+            $docPtr = Docblocks::getDocBlockPointer($phpcsFile, $stackPtr);
+            if (empty($docPtr)) {
+                $docPtr = $stackPtr;
+            }
+
+            $phpcsFile->addError(
+                'Missing @copyright tag',
+                $docPtr,
+                'CopyrightTagMissing'
+            );
+            return;
+        }
+    }
+
+    /**
+     * Process the file docblock and check for the presence of a @license tag.
+     *
+     * @param File $phpcsFile The file being scanned.
+     * @param int $stackPtr The position in the stack.
+     */
+    private function processFileLicense(File $phpcsFile, $stackPtr): void {
+        $tokens = $phpcsFile->getTokens();
+        $foundTokens = Docblocks::getMatchingDocTags($phpcsFile, $stackPtr, '@license');
+        if (empty($foundTokens)) {
+            $docPtr = Docblocks::getDocBlockPointer($phpcsFile, $stackPtr);
+            if ($docPtr) {
+                $phpcsFile->addError(
+                    'Missing @license tag',
+                    $docPtr,
+                    'LicenseTagMissing'
+                );
+            } else {
+                $phpcsFile->addError(
+                    'Missing @license tag',
+                    $stackPtr,
+                    'LicenseTagMissing'
+                );
+            }
+            return;
+        }
+
+        // If specified, get the regular expression from the config.
+        if (($regex = Config::getConfigData('moodleLicenseRegex')) !== null) {
```

Comment:
OT, and not needed for this PR, the other day I was reading that there is a way to document all these config options and also the configurable properties (in some XML format), that way when you look to the standard, all those options supported are shown.

I will create an issue about that once I find the details...

No action is needed here. Just sharing.

Reply:
About the documentation... the current Sniff documentation system does not support documenting sniff properties or configuration options, just found it (only describing the Sniff behaviour is allowed). So, f we want to document all the options... surely we'll need to invent something (structured) or go the manual way (to /docs).

Reference: https://github.com/PHPCSStandards/PHP_CodeSniffer/issues/148#issuecomment-2007469426

Ciao :)

Reply:
I was thinking and wondering the same. We haven't even got anything in the readme yet.